### PR TITLE
Ft2023 341: Adding code for custom form.

### PR DESCRIPTION
**Pull request by Souvik-Inno**
Added Custom config form-
 - Validated using ajax in a service class
 - Submitted the form and stored in settings file.

### CODE REVIEW

File: web/modules/custom/form_api/src/UserFormValidator.php [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+use Drupal\Core\Ajax\AddCssCommand;
+use Drupal\Core\Ajax\AjaxResponse;
+use Drupal\Core\Ajax\HtmlCommand;
+use Drupal\Core\Form\FormStateInterface;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Provides a service to validate user's form.
+ */
+class UserFormValidator {
```
Comment (vivekpan27) [inline web/modules/custom/form_api/src/UserFormValidator.php:15]: Why create a service instead of defining the methods in the form class itself?

Comment (Souvik-Inno) [reply web/modules/custom/form_api/src/UserFormValidator.php:15]: Sir isn't it better this way? As I was learning about separation of concerns, the rendering and submitting of form is done using the form class, so I thought to validate the form using this service which also includes a service to validate the email. Should I change it and do in the same class?

Comment (vivekpan27) [reply web/modules/custom/form_api/src/UserFormValidator.php:15]: Services are basically global functions which can be reused.
I don't think these methods you created can be reused in any other case.
At least for now.

Comment (Souvik-Inno) [reply web/modules/custom/form_api/src/UserFormValidator.php]: Okay. Have removed the service.

File: web/modules/custom/form_api/src/Form/UserForm.php [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+   *
+   * Display a tree of all the form elements of user form.
+   *
+   * @param array $form
+   *   An associative array containing the structure of the form.
+   * @param \Drupal\Core\Form\FormStateInterface $form_state
+   *   The current state of the form.
+   *
+   * @return array
+   *   The form structure.
```
Comment (vivekpan27) [inline web/modules/custom/form_api/src/Form/UserForm.php:89]: Do we need such extensive documentation for these inherited functions?

File: web/modules/custom/form_api/src/Form/UserForm.php [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+   * Validates the form using AJAX and submits it.
+   *
+   * @param array $form
+   *   An associative array containing the structure of the form.
+   * @param \Drupal\Core\Form\FormStateInterface $form_state
+   *   The current state of the form.
+   *
+   * @return \Drupal\Core\Ajax\AjaxResponse
+   *   The AJAX response.
+   */
```
Comment (vivekpan27) [inline web/modules/custom/form_api/src/Form/UserForm.php:166]: Same comment as above for this.

Comment (Souvik-Inno) [reply web/modules/custom/form_api/src/Form/UserForm.php:166]: The function submitUsingAjax is created by me and not an inherited function, it is called using ajax callback in $form['actions'].

File: web/modules/custom/form_api/src/Form/UserForm.php (hunk 2 above)
Comment (Souvik-Inno) [reply web/modules/custom/form_api/src/Form/UserForm.php]: Fixed.

**Comment by Souvik-Inno** [comment]
> Config schema is missing. Also default module configuration is missing. Please check how to add those.

Sir can you check once if files are ok?

File: config/sync/core.extension.yml [hunk 4]
```diff
@@ -7,7 +7,6 @@ module:
   automated_cron: 0
   big_pipe: 0
   block: 0
-  block_api: 0
```
Comment (vivekpan27) [inline config/sync/core.extension.yml:10]: Your PR should be clean.
I should not see any unrelated changes to the ticket.

File: web/modules/custom/form_api/src/install/form_api.settings.yml [hunk 5]
```diff
@@ -0,0 +1,4 @@
+full_name: 'Souvik Banerjee'
+phone_number: '+919873214560'
+email: this@gmail.com
+gender: male
```
Comment (vivekpan27) [inline web/modules/custom/form_api/src/install/form_api.settings.yml:4]: You can leave the values empty also.

File: web/modules/custom/form_api/src/Form/UserForm.php [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+    $email_value = $form_state->getValue('email');
+    $full_name = $form_state->getValue('full_name');
+    $email_validator = $this->emailValidator->isValid($email_value);
+    $email_providers = [
+      'gmail.com',
+      'yahoo.com',
+      'outlook.com',
+      'hotmail.com',
+      'mail.com',
+      'zoho.com',
```
Comment (vivekpan27) [inline web/modules/custom/form_api/src/Form/UserForm.php:210]: This should come from config.
What if tomorrow I want to add/remove from this list?
If this is part of code, then I have to raise PR, merge and wait for code to move to live.
If this is part of config, then I can easily edit the config in live using drush.

File: web/modules/custom/form_api/form_api.module [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+/**
+ * Implements hook_install().
+ */
+function form_api_install() {
+  $default_config = \Drupal::configFactory()->getEditable('form_api.settings');
+  $default_config->set('full_name', 'Souvik Banerjee');
+  $default_config->set('phone_number', '+919873214560');
+  $default_config->set('email', 'this@gmail.com');
+  $default_config->set('gender', 'male');
+  $default_config->save();
```
Comment (vivekpan27) [inline web/modules/custom/form_api/form_api.module:12]: Why add this?

File: web/modules/custom/form_api/form_api.module [hunk 8]
```diff
@@ -0,0 +1,13 @@
+<?php
+
+/**
+ * Implements hook_install().
+ */
+function form_api_install() {
+  $default_config = \Drupal::configFactory()->getEditable('form_api.settings');
```
Comment (vivekpan27) [inline web/modules/custom/form_api/form_api.module:7]: Did you test this?
Is this working for you?

**Review by vivekpan27 — CHANGES_REQUESTED** [review]

File: config/sync/core.extension.yml (hunk 4 above)
Comment (Souvik-Inno) [reply config/sync/core.extension.yml]: Okay done.

File: web/modules/custom/form_api/src/install/form_api.settings.yml (hunk 5 above)
Comment (Souvik-Inno) [reply web/modules/custom/form_api/src/install/form_api.settings.yml]: OKay, left empty.

File: web/modules/custom/form_api/src/Form/UserForm.php (hunk 6 above)
Comment (Souvik-Inno) [reply web/modules/custom/form_api/src/Form/UserForm.php]: Okay, implemented using config.

File: web/modules/custom/form_api/form_api.module (hunk 7 above)
Comment (Souvik-Inno) [reply web/modules/custom/form_api/form_api.module]: This was setting the config file on install using hook, now have removed it.

File: config/sync/core.extension.yml (hunk 4 above)
Comment (vivekpan27) [reply config/sync/core.extension.yml]: Now I do not see the entry for form_api module.

Comment (Souvik-Inno) [reply config/sync/core.extension.yml]: Okay, have done that now.

File: web/modules/custom/form_api/src/Form/UserForm.php [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+      '#required' => TRUE,
+      '#default_value' => $config->get('full_name') ?? '',
+      '#suffix' => "<div id='full-name-result' class='red'></div>",
+    ];
+    $form['phone_number'] = [
+      '#type' => 'tel',
+      '#title' => $this->t('Phone Number'),
+      '#description' => $this->t('Enter your phone number'),
+      '#required' => TRUE,
+      '#markup' => "<div id='phone-number-result'></div>",
```
Comment (vivekpan27) [inline web/modules/custom/form_api/src/Form/UserForm.php:95]: Can you put a screenshot for this?
I want to see how the markup looks.

**Review by vivekpan27 — APPROVED** [review]
